OpenRA.Game/Graphics/RgbaColorRenderer.cs

```diff
@@ -178,11 +192,19 @@ void DrawConnectedLine(float3[] points, float width, Color color, bool closed)
 		public void DrawLine(IEnumerable<float3> points, float width, Color color, bool connectSegments = false)
 		{
 			if (!connectSegments)
-				DrawDisconnectedLine(points, width, color);
+				DrawDisconnectedLine(points as float3[] ?? points.ToArray(), width, color);
```

Comment:
Wouldn't it be better (one enumeration rather than two, no extra allocation) to fall back to the `IEnumerable<float3>` version of `DrawDisconnectedLine` rather than converting it to an array?

Reply:
I copied it from the line below - DrawLineConnected. It should be changed there then also. I was not familiar with the as operator, i read it as check to see if it could be cast to float3.

Reply:
yeah we should change both places if my assumption about it being better is correct.

Reply:
The current implementation with the enumerator came from a previous round of optimizations to reduce memory pressure (see #10271), which the `.ToArray()` reverts.

A cleaner implementation might be to keep this line as it is, and instead change:
```csharp
void DrawDisconnectedLine(IEnumerable<float3> points, float width, Color color)
{
	// PERF: prefer direct enumeration for arrays
	var pointsArray = points as float3[];
	if (pointsArray != null)
	{
		var lastPoint = pointsArray[0];
		for (var i = 1; i < pointsArray.Length; i++)
		{
			var point = pointsArray[i];
			DrawLine(lastPoint, point, width, color);
			lastPoint = point;
		}
	}
	else
	{
		using (var e = points.GetEnumerator())
		{
			if (!e.MoveNext())
				return;

			var lastPoint = e.Current;
			while (e.MoveNext())
			{
				var point = e.Current;
				DrawLine(lastPoint, point, width, color);
				lastPoint = point;
			}
		}
	}
}
```

Reply:
If float3[] implements that interface. May cause the compiler to be able to optimize, 'see' the float3. Feel free to do as you see fit.

You may also want to implement the enumerated version of DrawConnectedLine, currently not available.